cmd/fluxd/main.go

```diff
@@ -106,106 +116,119 @@ func main() {
-		_                        = fs.Bool("k8s-in-cluster", true, "set this to true if fluxd is deployed as a container inside Kubernetes")
-		k8sSecretName            = fs.String("k8s-secret-name", "flux-git-deploy", "name of the k8s secret used to store the private SSH key")
-		k8sSecretVolumeMountPath = fs.String("k8s-secret-volume-mount-path", "/etc/fluxd/ssh", "mount location of the k8s secret storing the private SSH key")
-		k8sSecretDataKey         = fs.String("k8s-secret-data-key", "identity", "data key holding the private SSH key within the k8s secret")
-
-		// k8s-scope settings
-		k8sNamespaceWhitelist = fs.StringSlice("k8s-namespace-whitelist", []string{}, "restrict the view of the cluster to the namespaces listed. All namespaces are included if this is not set")
-		k8sAllowNamespace     = fs.StringSlice("k8s-allow-namespace", []string{}, "restrict all operations to the provided namespaces")
-		k8sDefaultNamespace   = fs.String("k8s-default-namespace", "", "the namespace to use for resources where a namespace is not specified")
-		k8sExcludeResource    = fs.StringSlice("k8s-unsafe-exclude-resource", []string{"*metrics.k8s.io/*", "webhook.certmanager.k8s.io/*", "v1/Event"}, "do not attempt to obtain cluster resources whose group/version/kind matches these glob expressions. Potentially unsafe, please read its documentation first")
-		k8sVerbosity          = fs.Int("k8s-verbosity", 0, "klog verbosity level")
+		versionFlag = fs.Bool("version", false, "get version number")
 
+		// not present in the config struct, since "extra for experts" or only relevant when running locally
+		kubernetesKubectl = fs.String("kubernetes-kubectl", "", "optional, explicit path to kubectl tool")
 		// SSH key generation
-		sshKeyBits   = optionalVar(fs, &ssh.KeyBitsValue{}, "ssh-keygen-bits", "-b argument to ssh-keygen (default unspecified)")
-		sshKeyType   = optionalVar(fs, &ssh.KeyTypeValue{}, "ssh-keygen-type", "-t argument to ssh-keygen (default unspecified)")
-		sshKeygenDir = fs.String("ssh-keygen-dir", "", "directory, ideally on a tmpfs volume, in which to generate new SSH keys when necessary")
-
-		// manifest generation
-		manifestGeneration = fs.Bool("manifest-generation", false, "search for .flux.yaml files to generate manifests")
-
-		// upstream connection settings
-		upstreamURL = fs.String("connect", "", "connect to an upstream service e.g., Weave Cloud, at this base address")
-		token       = fs.String("token", "", "authentication token for upstream service")
-		rpcTimeout  = fs.Duration("rpc-timeout", 10*time.Second, "maximum time an operation requested by the upstream may take")
-
-		dockerConfig = fs.String("docker-config", "", "path to a docker config to use for image registry credentials")
+		sshKeyBits = optionalVar(fs, &ssh.KeyBitsValue{}, "ssh-keygen-bits", "-b argument to ssh-keygen (default unspecified)")
+		sshKeyType = optionalVar(fs, &ssh.KeyTypeValue{}, "ssh-keygen-type", "-t argument to ssh-keygen (default unspecified)")
```

Comment:
Why doesn't it make sense to put these two in the config file?

Reply:
They've always been off to the side of the "normal" flags, perhaps since they pertain to bootstrapping Flux rather than how it operates once running, or because the values are sent straight to the `ssh` package (and not examined). I don't think either of those are a particularly strong rationale; on the other hand, it would take some reimplementation to be able to put them in the config file, and I don't think it's worth it.

Reply:
I see. IMO they should either be allowed in the configfile or deprecated. I think it's important to keep the invariant of "everything can be configured through both flags and the config file" for currently non-deprecated flags.

Reply:
> I think it's important to keep the invariant of "everything can be configured through both flags and the config file" for currently non-deprecated flags.

I don't. For example, `--kubernetes-kubectl` is a flag that is useful (well, kind of), but doesn't make sense in the config file, since it is about the environment in which fluxd runs rather than its behaviour when running.

Actually, a better example: `--k8s-secret-volume-mount-path` only exists to tell fluxd about where a volume has been mounted for it to use, i.e., you match its argument with a mount path elsewhere in the same Deployment definition. Putting it in the config file makes that into a trapeze act -- now, the values in two separate files, possibly from two completely different sources, must match, and there's no reason to vary them independently. (And I should really take that field out of the config file definition, now I have noticed it.)

Reply:
`--docker-config` arguably belongs in the same category; useful only when it matches something specified elsewhere in the Deployment manifest (or when running locally).